### PR TITLE
Replace deprecated tempdir/atomicwrites dependencies

### PR DESCRIPTION
**Pull request by kornelski**
The `tempdir` crate is dead.

### CODE REVIEW

**Comment by kornelski** [comment]
@fulmicoton what do you think about the failure of atomic write test? I'm not sure if that's due to difference of implementation of atomicwrite vs tempfile: https://github.com/Stebalien/tempfile/pull/106 — or just a bug elsewhere?

**Comment by fulmicoton** [comment]
@kornelski I don't understand. It looks like you replaced `AtomicWrite` by named tempfile. Are Tempfile supposed to offer the same guarantee to begin with?

File: src/directory/mmap_directory.rs [hunk 1]
```diff
@@ -506,8 +506,9 @@ impl Directory for MmapDirectory {
     fn atomic_write(&mut self, path: &Path, data: &[u8]) -> io::Result<()> {
         debug!("Atomic Write {:?}", path);
         let full_path = self.resolve_path(path);
-        let meta_file = atomicwrites::AtomicFile::new(full_path, atomicwrites::AllowOverwrite);
-        meta_file.write(|f| f.write_all(data))?;
+        let mut temp = NamedTempFile::new_in(full_path.parent().expect("has filename"))?;
```
Comment (fulmicoton) [inline src/directory/mmap_directory.rs:509]: This looks like a bug to me.

**Comment by kornelski** [comment]
https://lib.rs/atomicwrites readme says:

> Alternatives
>
> tempfile has a persist method doing the same thing.

And atomicwrites is implemented using the `tempdir` crate:

https://github.com/untitaker/rust-atomicwrites/blob/a0db3c6ad80b834c50ee26fda51afd8c3d3f24d0/src/lib.rs#L126-L133

but `tempdir` is deprecated, obsoleted by `tempfile`.

So effectively both do exactly the same thing - make a named temp file, write to it, then rename over original. The only difference I could see is that atomicwrites sets `MOVEFILE_WRITE_THROUGH` on windows, and tempfile doesn't.

**Comment by fulmicoton** [comment]
@kornelski Ah I see! The unit test failing is related to the way we watch the directory. (using the notify crate). It has some buggy behavior (starting to watch triggered some error on some plateforms), so I actually intend to deprecate it.

I think we can merge your branch once this is done. This is tracked by #602.

**Comment by fulmicoton** [comment]
@kornelski thanks for the research and the analysis. This is of great help!

**Comment by fulmicoton** [comment]
@kornelski should we close this then?

**Comment by kornelski** [comment]
This removes atomicwrites, which removes tempdir from dependencies, so I still think it's worth doing. 

As far as I understand it's waiting for #602 and/or `directory::tests::test_mmap_directory` to be changed.

**Pull request closed without merging**